server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/processor/AstProcessor.java

```diff
@@ -74,20 +77,25 @@ public void process(ProcessingPhase phase, Node node, ProcessingContext ctx) {
   /**
    * Process tree node and its children after tree construction.
    *
-   * @param processor list of available processors
+   * @param processors list of available processors
    * @param node a node to process
    * @param ctx processing context
    */
   private void process(Map<Class<? extends Node>, List<Processor<? extends Node>>> processors,
       Node node, ProcessingContext ctx) {
     ThreadInterruptionUtil.checkThreadInterrupted();
     final Class<? extends Node> nodeClass = node.getClass();
-    processors.forEach((key, value) -> {
-      if (!key.isAssignableFrom(nodeClass))
-        return;
-      for (Processor<? extends Node> p : value)
-        ((Processor<Node>) p).accept(node, ctx);
-    });
-    node.getChildren().forEach(n -> process(processors, n, ctx));
+    if (nodeClass == ProgramNode.class) {
+      ctx.getCurrentProgramNodeStack().push((ProgramNode) node);
+    }
+    for (Processor<? extends Node> processor : processors.getOrDefault(nodeClass, Collections.emptyList())) {
+      ((Processor<Node>) processor).accept(node, ctx);
+    }
+    for (Node n : node.getChildren()) {
+        process(processors, n, ctx);
+    }
+    if (nodeClass == ProgramEndNode.class) {
```

Comment:
Should this be done in a finally block?

Reply:
Yes, but with check for ProgramNode instead.